### ISSUE TITLE
evaluation of multivariate polynomials

### ISSUE THREAD

**Issue by heiderich** [Nemocas/AbstractAlgebra.jl#106]
Is there a particular reason why multivariate polynomials can only be evaluated at values that are elements of their parent ring?

In my opinion it would be desirable if one could evaluate them at values belonging to any commutative algebra over the parent ring.

I think this would require a data structures for algebras. Does it exist or is it supposed to be defined?

**Comment by fieker** [comment]
On Fri, Jun 22, 2018 at 09:40:20AM +0000, Florian Heiderich wrote:
> Is there a particular reason why multivariate polynomials can only be evaluated at values that are elements of their parent ring?
> 
> In my opinion it would be desirable if one could evaluate them at values belonging to any commutative algebra over the parent ring.
> 
> I think this would require a data structures for algebras. Does it exist or is it supposed to be defined?
> 
Two different problems
 - I agree you should be able to evaluate more freely
   Possibly using a range of methods depending on the application.
   Possibly this should also be tied to the maps (or can be tied to the
   maps)
 - Algebras: those are starting to exist in Hecke and they (implicitly)
   exist in Singular. It is not quite clear what one can do generically
   and thus what should be done in AbstractAlgebra
   Do you have any proposals?

**Comment by thofma** [comment]
I don't think it is necessary to have the data structure for this problem.

The evaluation functions needs only small changes, then it works for all ring elements. The heavy lifting is done by univariate evaluation, for which there are no restrictions if I remember correctly. I think only the special return values have to be adjusted to return `zero(base_ring(A[1]))` instead of `zero(base_ring(a))`.

**Comment by heiderich** [comment]
> The evaluation function needs only small changes, then it works for all ring elements. 

@thofma: What do you mean by "all ring elements"? As far as commutative rings are concerned, I only know how to evaluate polynomials in the following situation:

Say p is an element of a polynomial ring over a commutative base ring R in n variables.
If S is a commutative R-algebra and (s_1, ..., s_n) an n-tuple of elements of S, then you can define p(s_1, ..., s_n). In fact you have a R-algebra homomorphism from R[x_1, ..., x_n] to S. This is just the universal property of polynomial rings.

As far as commutative rings are concerned, I would imagine that a data structure for an R-algebra could be a pair (S,f) consisting of a commutative ring S together with a ring homomorphism f from R to S. This however would first require a data structure for a ring homomorphism. I guess one should have an interface for ring homomorphisms and then possibly allow different implementations.

**Comment by thofma** [comment]
Yes, I meant all commutative R-algebras S. Practically in AbstractAlgebra this just means that r*s should work, where r is in R and s in S.

I just want to say that we don't need another type. Just relax the signature and adjust the return values.

**Comment by heiderich** [comment]
Yes, from a technical point of view this is sufficient as long as the rings are commutative.

Is there a consensus that one should just relax the signature and leave it to the user to call this function only if this multiplication is possible?

**Comment by thofma** [comment]
At least this is what is happening for univariate polynomials:

```julia
julia> Qx, x = QQ["x"]
(Univariate Polynomial Ring in x over Rationals, Fx)

julia> F = ResidueField(ZZ, 5)
Residue field of Integers modulo 5

julia> evaluate(x^2, F(1))
ERROR: MethodError: no method matching (::AbstractAlgebra.Integers{BigInt})(::Rational{BigInt})
Stacktrace:
 [1] *(::Rational{BigInt}, ::AbstractAlgebra.Generic.ResF{BigInt}) at /home/thofmann/.julia/v0.6/AbstractAlgebra/src/generic/ResidueField.jl:252
 [2] evaluate(::AbstractAlgebra.Generic.Poly{Rational{BigInt}}, ::AbstractAlgebra.Generic.ResF{BigInt}) at /home/thofmann/.julia/v0.6/AbstractAlgebra/src/generic/Poly.jl:1575

julia> Zx, x = ZZ["x"]
(Univariate Polynomial Ring in x over Integers, x)

julia> evaluate(x^2, F(1))
1
```

**Comment by heiderich** [comment]
Just changing the signature and the return type seems to lead to an infinite recursion.

**Comment by thofma** [comment]
Sorry, my bad. I had some misunderstanding about the algorithm implemented.

To use the current implementation, I think one could `f` to be a polynomial of `S`, where `S` is the ring containing elements one wants to evaluate `f` at. So something like `evaluate(change_ring(f, S), A)`. Unfortunately we don't have a change ring function. Also I don't know if this is the best approach since one has to create a new multivariate ring for every evaluation.

**Comment by thofma** [comment]
I will take only about commutative rings in the following. There should be `change_ring(f::MPoly, g)` where `g` is anything callable which turns elements of `R` into elements of `S`. Can be a ring morphism but it need not be. The function will just apply `g` to all the coefficents of `f`. Then `change_ring(f, S)` can fall back to `change_ring(f, a -> S(a))` or `change_ring(f, a -> one(S)*a)`. Or, if you have a structure morphism `g` you can just call `change_ring(f, g)`.

Here are my reasons why I want to keep it simple: One should be able to easily turn polynomials over Z into polynomials Q. For this I don't want to define a new object with a new type, just to represent Q as a Z-algebra. Also in applications I want to go from polynomials over Q to polynomials over Z/pZ, with the possibily of failure in case the denominator is not invertible modulo p.

**Comment by heiderich** [comment]
I added such a function `change_base_ring(f::MPoly,g)` in pull request #109. I do not consider it do be nice, but it seems to fulfill what you propose.
